sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -1032,7 +1026,11 @@ class SimpleTableFunctionRegistry extends SimpleFunctionRegistryBase[LogicalPlan
 
   override def clone(): SimpleTableFunctionRegistry = synchronized {
```

Comment:
I think there is a problem here. We don't synchronize the write path, how can we safely clone the `ConcurrentHashMap`?

Reply:
We can introduce `ReentrantReadWriteLock` guarding the `ConcurrentHashMap`.
`clone` and `clear` would take the write lock on the `ConcurrentHashMap`.
The other methods take read lock on the map.

Reply:
It may work, but makes the code complicated. We should only do it if this does make a difference to real-world workloads.

Reply:
In fact, It's safe here.
Because the `entrySet()` of `ConcurrentHashMap` is thread safe. We don't need synchronized.